### ISSUE TITLE
functioning of sample constraints

### ISSUE THREAD

**Issue by ciri1998** [benmoseley/FBPINNs#11]
Good morning, I'm not sure I fully understand how "sample_constraints" works. For example in the WaveEquationConstantVelocity3D problem, if I understand correctly:
(0,(0,0)) the first zero indicates u and the other two the second derivative in t.
While (0,(1,1)) and (0,(2,2)) are the second derivative of u in x and y respectively. it's correct?
Now my problem arises because I would like to determine the individual components of u along x and y. For example, if u is a velocity, I need the gradients of the x component and the y component to appear in the loss fn, for example:
phys = u_x + d(u_x)/dx + d(u_x)/dy + d(u_y)/dy + u_y
with u_x the x component of u and u_y the y component.
it is possible to determine them by going to edit
(0,(1,))
(1,(0,))
(0,(0,))
(1,(1,))
or similarly?

**Comment by benmoseley** [comment]
yes, for example (0,(1,)) gives $\frac{\partial u_0}{\partial x_1}$, (1,(0,)) gives $\frac{\partial u_1}{\partial x_0}$, (1,(0,1)) gives $\frac{\partial^2 u_1}{\partial x_0 \partial x_1}$ etc, if modelling a vector function $\mathbf{u}(\mathbf{x})$

**Comment by ciri1998** [comment]
thanks for your answer.
I have some doubts regarding data loss.
For example, in sample_constraint i write:
 # loss of physics
 x_batch_phys = domain.sample_interior(all_params, key, sampler, batch_shapes[0])
 required_ujs_phys = (
 (0,()), # u
 (1,()), # see
 (0,(0,)), # ux = du/dx
 (0,(1,)), # uy = du/dy
 (1,(0,)), # vx
 (1,(1,)), # vy
 )
for physics loss, and:
# data loss
 x_dat= jnp.array(np.linspace(-1.2,2.8,50).astype(float)).reshape(50,1)
 y_dat= jnp.array(np.linspace(-1,1,50).astype(float).reshape(50,1))

 X, Y = np.meshgrid(x_dat, y_dat)


 x_batch_data = jnp.column_stack([X.ravel(), Y.ravel()])


 batch_shape=c.n_test

 u_data = WaveEquation3D.exact_solution(all_params, x_batch_data, batch_shape[0])
 requested_ujs_data = (
 (0,()),
 (1,()),
 )
 return [[x_batch_phys, required_ujs_phys],[x_batch_data, u_data, required_ujs_data]]

inside the loss_fn:

 x_batch, u, v, ux, uy, vx, vy = constraints[0]

for physics loss, and:
 uc, vc, u, v = constraints[1]

for data loss, correct?
because as far as the physics loss is concerned the results seem to make sense, but when I add the data loss the convergence slows down a lot and always remains almost meaningless. I guess it may come from a data loss error in loss_fn
